core/server/worker/src/main/java/alluxio/worker/block/BlockLockManager.java

```diff
@@ -126,6 +140,12 @@ private OptionalLong lockBlockInternal(long sessionId, long blockId, BlockLockTy
     ClientRWLock blockLock = getBlockLock(blockId);
     Lock lock = blockLockType == BlockLockType.READ ? blockLock.readLock() : blockLock.writeLock();
     // Make sure the session isn't already holding the block lock.
+    // todo(bowen): This is a best-effort check and is subject to race condition.
+    // Remove this check and implement better error signaling or retry.
+    // When the said race condition occurs, the write lock will block indefinitely until the
+    // reader releases the lock. In case the reader panics while holding the lock, this will lead
+    // to a deadlock. This can happen e.g. when worker tries to move a block to a new location while
+    // a client is actively reading it.
     if (blockLockType == BlockLockType.WRITE && sessionHoldsLock(sessionId, blockId)) {
```

Comment:
I think it's better we keep this check because a UT is specifically testing it: `alluxio.worker.block.BlockLockManagerTest#lockAlreadyReadLockedBlock`. Though the original implementation is subject to race condition as well.